### PR TITLE
New Diacritics Component for Inputs

### LINKED ISSUE #1805: Use MediaWiki JSON to create a component for entering diacritics
**Describe the feature you'd like**
In Mediawiki (the wiki software behind Wikimedia) when the user is entering text in their rich text fields, the user can expand and display a list of special characters for a number of different character sets that allows them to enter a diacritic character by clicking on the the character that is then inserted into the text field.

Screen Shot 2019-11-19 at 12 51 52 PM

This widget is being populated from a JSON file in the Mediawiki PHP code base that we could use to base this component on:

https://raw.githubusercontent.com/wikimedia/mediawiki/master/resources/src/mediawiki.language/specialcharacters.json

**Describe alternatives you've considered**
Catalogers can copy and paste diacritics from other software but this isn't the best option, especially for non-English cataloging.

### PR DESCRIPTION
**Pull request by jermnelson**
Displays a list of vocabularies that a user can select and then click on buttons to enter special characters into the input field.

**Note** Only the `InputLiteral` component is currently supported, the `InputLookup` based components will not work with the current strategy. 

diacritics-selection

Fixes #1805

### CODE REVIEW

File: src/components/editor/diacritics/CharacterButton.jsx [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+  const handleClick = (event) => {
+    const existingValue = targetRef.current.value
+    targetRef.current.value = existingValue + props.character
+    targetRef.current.focus()
+    event.preventDefault()
+  }
+
+  return (
+    <button className="btn btn-light"
+            style={{ margin: '.25em' }}
```
Comment (justinlittman) [inline src/components/editor/diacritics/CharacterButton.jsx]: Most of our other margins are in pixels. Can you switch?

File: static/specialcharacters.json [hunk 2]
```diff
@@ -0,0 +1,184 @@
+{
```
Comment (justinlittman) [inline static/specialcharacters.json]: In the README, or elsewhere can you describe where this file comes from and/or if it needs to be updated?

File: src/store.js [hunk 3]
```diff
@@ -12,7 +12,12 @@ export const initialState = {
       lastChecked: Date.now(),
     },
     editor: { // The state of the editor
-      currentResource: undefined,
+      currentPath: undefined,
+      diacritics: {
+        show: false,
+        targetRef: undefined,
```
Comment (justinlittman) [inline src/store.js]: Can we avoid storing a ref and function in state? We don't do this anywhere else.

File: src/components/editor/property/InputLiteral.jsx [hunk 4]
```diff
@@ -109,17 +137,19 @@ const mapStateToProps = (state, ownProps) => {
   const displayValidations = getDisplayResourceValidations(state)
   const formData = findNode(state, reduxPath)
   const errors = findResourceValidationErrorsByPath(state, reduxPath)
+  const diacriticMode = state.selectorReducer.editor.diacritics.show
```
Comment (justinlittman) [inline src/components/editor/property/InputLiteral.jsx]: Calling this diacriticMode when a boolean is confusing. Perhaps shouldShowDiacritic or displayDiacritic?

File: src/components/editor/property/InputLiteral.jsx [hunk 5]
```diff
@@ -58,6 +63,15 @@ const InputLiteral = (props) => {
     inputLiteralRef.current.focus()
   }
 
+  const handleDiacritics = (event) => {
```
Comment (justinlittman) [inline src/components/editor/property/InputLiteral.jsx]: toggleDiacritics would be clearer.

File: src/components/editor/diacritics/VocabChoice.jsx [hunk 6]
```diff
@@ -0,0 +1,23 @@
+// Copyright 2019 Stanford University see LICENSE for license
+
+import React from 'react'
+import PropTypes from 'prop-types'
+import specialcharacters from '../../../../static/specialcharacters.json'
+
+const VocabChoice = (props) => {
+  const getOptions = () => {
+    const options = []
+    Object.keys(specialcharacters).forEach((key) => {
```
Comment (justinlittman) [inline src/components/editor/diacritics/VocabChoice.jsx]: Slightly cleaner as a map.

File: src/components/editor/diacritics/VocabChoice.jsx [hunk 7]
```diff
@@ -0,0 +1,23 @@
+// Copyright 2019 Stanford University see LICENSE for license
+
+import React from 'react'
+import PropTypes from 'prop-types'
+import specialcharacters from '../../../../static/specialcharacters.json'
```
Comment (justinlittman) [inline src/components/editor/diacritics/VocabChoice.jsx]: Can this just be `static/specialCharacters`?

File: src/components/editor/diacritics/DiacriticsSelection.jsx [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+    cssClasses = ['sticky-top', 'bg-white']
+  } else {
+    cssClasses = ['d-none']
+  }
+
+  const selectVocabulary = (event) => {
+    const vocabChoice = event.target.value
+    if (vocabChoice === undefined) return null
+    const buttons = []
+    specialcharacters[vocabChoice].characters.forEach((char) => {
```
Comment (justinlittman) [inline src/components/editor/diacritics/DiacriticsSelection.jsx]: Slightly cleaner as a map.

File: src/components/editor/diacritics/DiacriticsSelection.jsx [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+      buttons.push(<CharacterButton key={shortid.generate()}
```
Comment (justinlittman) [inline src/components/editor/diacritics/DiacriticsSelection.jsx]: Why not just use the character as the key?

File: src/components/editor/diacritics/CharacterButton.jsx (hunk 1 above)
Comment (jermnelson) [reply src/components/editor/diacritics/CharacterButton.jsx]: Done.

File: static/specialcharacters.json (hunk 2 above)
Comment (jermnelson) [reply static/specialcharacters.json]: Done, added to README

File: src/components/editor/property/InputLiteral.jsx (hunk 4 above)
Comment (jermnelson) [reply src/components/editor/property/InputLiteral.jsx]: Done, renamed to `shouldShowDiacritics`.

File: src/components/editor/property/InputLiteral.jsx (hunk 5 above)
Comment (jermnelson) [reply src/components/editor/property/InputLiteral.jsx]: Done.

File: src/components/editor/diacritics/VocabChoice.jsx (hunk 6 above)
Comment (jermnelson) [reply src/components/editor/diacritics/VocabChoice.jsx]: Done.

File: src/components/editor/diacritics/VocabChoice.jsx (hunk 7 above)
Comment (jermnelson) [reply src/components/editor/diacritics/VocabChoice.jsx]: I'm not able to get `static/specialcharacters.json` to compile.

File: src/components/editor/diacritics/DiacriticsSelection.jsx (hunk 8 above)
Comment (jermnelson) [reply src/components/editor/diacritics/DiacriticsSelection.jsx]: Done

File: src/components/editor/diacritics/DiacriticsSelection.jsx (hunk 9 above)
Comment (jermnelson) [reply src/components/editor/diacritics/DiacriticsSelection.jsx]: In some of the vocabularies, the character is not unique.

File: src/store.js (hunk 3 above)
Comment (jermnelson) [reply src/store.js]: Done, now passing reduxState and no longer trying to set the language code based on the character set.

File: README.md [hunk 10]
```diff
@@ -46,6 +46,8 @@ You will need to be online to authenticate even when developing in localhost.
 - See `package.json` for npm package dependencies.
 - The web server used is the `express` web framework for node.js
 - React components are located in `src/components/` directory
+- Characters for the diacritics components come from the
+  [Mediawiki](https://github.com/wikimedia/mediawiki/) repository's [specialcharacters.json](https://github.com/wikimedia/mediawiki/blob/master/resources/src/mediawiki.language/specialcharacters.json). This file has been customized for use in Sinopia.
```
Comment (justinlittman) [inline README.md]: Can you describe what you mean by "customized"?

File: src/store.js (hunk 3 above)
Comment (justinlittman) [inline src/store.js]: I think the default state needs to be fixed.

File: src/components/editor/diacritics/DiacriticsSelection.jsx [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+
+  const selectVocabulary = (event) => {
+    const vocabChoice = event.target.value
+    if (vocabChoice === undefined) return null
+    const buttons = []
+    specialcharacters[vocabChoice].characters.map((char) => {
+      buttons.push(<CharacterButton key={shortid.generate()}
+                                    character={char} />)
+    })
+    dispatch(setDiacriticsButtons(buttons))
```
Comment (justinlittman) [inline src/components/editor/diacritics/DiacriticsSelection.jsx]: As discussed, would prefer not to store this in redux state.

File: README.md (hunk 10 above)
Comment (jermnelson) [reply README.md]: Done, added more context to the changes.

File: src/components/editor/diacritics/DiacriticsSelection.jsx (hunk 11 above)
Comment (jermnelson) [reply src/components/editor/diacritics/DiacriticsSelection.jsx]: Done, reverted to using local component state for character buttons. **Note** this means that we preserve the state of the `DiacriticsSelection` (buttons and selected character set) between invocations by the user.

File: src/store.js [hunk 12]
```diff
@@ -12,7 +12,11 @@ export const initialState = {
       lastChecked: Date.now(),
     },
     editor: { // The state of the editor
-      currentResource: undefined,
+      currentPath: undefined,
```
Comment (justinlittman) [inline src/store.js]: Was this supposed to change?

File: src/components/editor/property/InputLiteral.jsx [hunk 13]
```diff
@@ -74,19 +88,30 @@ const InputLiteral = (props) => {
...
+              required={required}
+              className="form-control"
+              placeholder={props.propertyTemplate.propertyLabel}
+              onChange={event => setContent(event.target.value)}
+              onKeyPress={handleKeypress}
+              onBlur={handleBlur}
+              value={content}
+              disabled={disabled}
+              ref={inputLiteralRef}
+              id={props.reduxPath.join('')}
```
Comment (justinlittman) [inline src/components/editor/property/InputLiteral.jsx]: Would prefer including a prefix (e.g., `literal-`) to avoid id clashes.

**Comment by justinlittman** [comment]
Another gotcha. Steps to reproduce:
1. Open an Abbreviated title.
2. Click +Add.
3. Click A-umlat to open the keyboard.
4. Select a character set and click on some characters.
5. Click X to close the keyboard.
6. Click Edit next to the entered characters.
7. Click inside the input box.
8. Click A-umlat to open the keyboad.

Expected result:
The keyboard opens.

Actual result:
It exits edit mode (the characters are blue and the input is cleared).

File: src/store.js (hunk 12 above)
Comment (jermnelson) [reply src/store.js]: Nope, now fixed

File: src/components/editor/property/InputLiteral.jsx (hunk 13 above)
Comment (jermnelson) [reply src/components/editor/property/InputLiteral.jsx]: Done.

**Comment by jermnelson** [comment]
> Tiny tweaks and a gotcha.
> 
> Steps to reproduce:
> 
>     1. Open an Abbreviated title.
> 
>     2. Click +Add.
> 
>     3. Click A-umlat to open the keyboard.
> 
>     4. Select the character set and enter some characters.
> 
>     5. Click X to close the keyboard.
> 
>     6. Click Edit next to the characters.
> 
>     7. Click A-umlat to open the keyboard.
> 
>     8. Enter some more characters.
> 
>     9. Click X to close the keyboard.
> 
> 
> Expected result:
> Characters appear in blue and input is cleared.
> 
> Actual result:
> Characters appear in blue and input contains the originally entered characters.
> image

This is now fixed with the latest commit.

**Comment by justinlittman** [comment]
Verified that is fixed.

**Pull request merged**